### DISCUSSION TITLE
Delivery Semantics & Failed message handling

### DISCUSSION THREAD

**Discussion by FabianTaggr** [rabbitmq/rabbitmq-stream-dotnet-client#397] [Q&A]
Hello! Excited about this project.

But - there's one thing i cannot wrap my head around. When handling of a message fails on the consumer side, why doesn't this "poison" the consumer? As of now i just moves on the next message, which is quite problematic if you ask me. It means I can't really guarantee a message has been successfully processed my application, as it implicitly just skips failed ones. 

See my reference implementation of the message handler:

```
MessageHandler = async (_, consumer, context, message) =>
            {
                try
                {
       
                    throw new Exception();            
                }
                catch (Exception ex)
                {
                    _logger!.LogError(ex, "Error handling message");
                    throw;
                }
           }
```

Should i just infinitely retry to essentially block the thread from consuming the next message?

**Comment by Gsantomaggio** [comment]
hi @FabianTaggr ,
Thank you for your feedback :) 
I am trying to understand. What do you think should happen?


 If you want to close the consumer in case of exception, it is enough to call `consumer.close()` in case of error:

```csharp
MessageHandler = async (_, consumer, context, message) =>
            {
                try
                {
       
                    throw new Exception();            
                }
                catch (Exception ex)
                {
                    _logger!.LogError(ex, "Error handling message");
consumer.close(); ///<<<<<<---------here
                }
           }
```

**Comment by Gsantomaggio** [comment]
Will close due to inactivity.
